gql/type_adaptor.py

```diff
@@ -0,0 +1,123 @@
+from typing import Any, Dict, Callable, Optional, List
+
+from graphql.type.schema import GraphQLSchema
+from graphql.type.definition import GraphQLObjectType, GraphQLField, GraphQLScalarType
+
+
+class TypeAdaptor(object):
+    """Substitute custom scalars in a GQL response with their decoded counterparts.
+
+    GQL custom scalar types are defined on the GQL schema and are used to represent
+    fields which have special behaviour. To define custom scalar type, you need
+    the type name, and a class which has a class method called `parse_value()` -
+    this is the function which will be used to deserialize the custom scalar field.
+
+    We first need iterate over all the fields in the response (which is done in
+    the `_traverse()` function).
+
+    Each time we find a field which is a custom scalar (it's type name appears
+    as a key in self.custom_scalars), we replace the value of that field with the
+    decoded value. All of this logic happens in `_substitute()`.
+
+    Public Interface:
+    apply(): pass in a GQL response to replace all instances of custom
+        scalar strings with their deserialized representation."""
+
+    def __init__(self, schema: GraphQLSchema, custom_scalars: Dict[str, Any] = {}) -> None:
+        """ schema: a graphQL schema in the GraphQLSchema format
+            custom_scalars: a Dict[str, Any],
+                where str is the name of the custom scalar type, and
+                      Any is a class which has a `parse_value()` function"""
```

Comment:
What's the signature of `parse_value`? It may be worth creating an abstract class that we expect the values in `custom_scalars` to subclass.

Reply:
At least in our case that would be a little awkward since (for example) we already have a fleshed out CurrencyAmount class. It would mean that the class would inherit from multiple base classes and CurrencyAmount would be tightly coupled to the gql client, which seems like it is more trouble than it is worth.

I will update the comment to make it clear what the signature of `parse_value` is